freenas/cli/plugins/openvpn.py

```diff
@@ -0,0 +1,473 @@
+            get='ca',
+            set='ca',
+            type=ValueType.STRING,
+            usage=_('''\
+            PKI mode server CA'''),
+        )
+
+@description("Allows to bridge openvpn interface to the main interface")
+class OpenVPNBridgeCommand(Command):
+    """
+    Usage: bridge
+
+    Allows to bridge openvpn interface to the main interface.
+    This property is only allowed in pki mode.
+
+    """
+    def run(self, context, args, kwargs, opargs):
+
+        vpn_confg = context.call_sync('service.openvpn.get_config')
+        if vpn_confg['mode'] == 'psk':
+            raise CommandException(_('Bridging to main interface is possible only in pki mode.'))
+
+        context.submit_task('service.openvpn.bridge')
+
+@description("Allows to generate OpenVPN cryptographic properties")
+class OpenVPNCryptoCommand(Command):
+    """
+    Usage: generate_crypto key_type=[tls-auth-key or dh-parameters] key_length=[1024 or 2048] bits
+    Example: generate_crypto key_type=tls-auth-key
+             generate_crypto key_type=dh-parameters key_length=2048
+
+    Creates OpenVPN cryptographic Values.
+    dh-parameters can be either 1024 or 2048 bits long
+    tls-auth-key is always generated as 2048 bit value
+    """
+    def run(self, context, args, kwargs, opargs):
+
+        if len(kwargs) < 1:
+            raise CommandException(_("generate_cryto requires more arguments, see 'help generate_cryto' for more information"))
+
+        if not kwargs['key_type'] or kwargs['key_type'] not in ['dh-parameters', 'tls-auth-key']:
+            raise CommandException(_("wrong arguments, see 'help generate_cryto' for more information"))
+
+        if kwargs['key_type'] == 'dh-parameters' and kwargs['key_length'] not in [1024, 2048]:
+            raise CommandException(_("wrong key_length value, see 'help generate_cryto' for more information"))
+
+        context.submit_task('service.openvpn.gen_key', kwargs['key_type'], kwargs.get('key_length'))
+
+@description("Provides corresponding OpenVPN client configuration")
+class OpenVPNClientConfigCommand(Command):
+    """
+    Usage: provide_client_config
+
+    Provides corresponding OpenVPN client configuration.
+    It needs to be copied to the OpenVPN client machine.
+    """
+    def run(self, context, args, kwargs, opargs):
+
+        vpn_client_confg = context.call_sync('service.openvpn.client_config.provide_config')
+        return Sequence(vpn_client_confg)
+
+@description("Provides OpenVPN static key")
+class OpenVPNStaticKeyCommand(Command):
+    """
+    Usage: provide_static_key
+
+    Provides static OpenVPN key that needs to be copied to the client machine
+    """
+    def run(self, context, args, kwargs, opargs):
+
+        vpn_client_confg = context.call_sync('service.openvpn.get_config')
+        return Sequence(vpn_client_confg['tls_auth'])
+
+
+@description("Manages OpenVPN service")
+class OpenVPNManageNamespace(EntitySubscriberBasedLoadMixin, EntityNamespace):
```

Comment:
That namespace doesn't quite make sense - it could show only one thing at most. Why does it need to be in separate namespace anyways? Can't one go to `service openvpn` to do start/stop/restart/enable/disable?